Marlin/src/feature/bedlevel/bdl/bdl.cpp

```diff
@@ -42,151 +43,199 @@
-    static float old_cur_z = cur_z,
-                 old_buf_z = current_position.z;
-
+  //if (config_state == BDS_IDLE) return;
+  static millis_t next_check_ms = 0; // starting at T=0
+  static float zpos = 0.0f;
+  const millis_t ms = millis();
+  if (ELAPSED(ms, next_check_ms)) { // timed out (or first run)
+    next_check_ms = ms + (config_state < 0 ? 200 : 50);   // check at 5Hz or 20Hz
+
+    uint16_t tmp = 0;
+    const float cur_z = planner.get_axis_position_mm(Z_AXIS) - pos_zero_offset;
+    static float old_cur_z = cur_z, old_buf_z = current_position.z;
     tmp = BD_I2C_SENSOR.BD_i2c_read();
-    if (BD_I2C_SENSOR.BD_Check_OddEven(tmp) && (tmp & 0x3FF) < 1020) {
-      const float z_sensor = (tmp & 0x3FF) / 100.0f;
-      if (cur_z < 0) config_state = 0;
-      //float abs_z = current_position.z > cur_z ? (current_position.z - cur_z) : (cur_z - current_position.z);
+    if (BD_I2C_SENSOR.BD_Check_OddEven(tmp) && good_data(tmp)) {
+      const float z_sensor = interpret(tmp);
+      if (cur_z < -0.5f) config_state = BDS_IDLE;
+      //const float abs_z = ABS(current_position.z - cur_z);
       #if ENABLED(BABYSTEPPING)
-        if (cur_z < config_state * 0.1f
-          && config_state > 0
-          && old_cur_z == cur_z
-          && old_buf_z == current_position.z
-          && z_sensor < (MAX_BD_HEIGHT)
-        ) {
-          babystep.set_mm(Z_AXIS, cur_z - z_sensor);
-          #if ENABLED(DEBUG_OUT_BD)
-            SERIAL_ECHOLNPGM("BD:", z_sensor, ", Z:", cur_z, "|", current_position.z);
-          #endif
-        }
-        else {
-          babystep.set_mm(Z_AXIS, 0);          //if (old_cur_z <= cur_z) Z_DIR_WRITE(HIGH);
-          stepper.apply_directions();
+        if (config_state > 0) {
+          if (cur_z < config_state * 0.1f
+            && old_cur_z == cur_z
+            && old_buf_z == current_position.z
+            && z_sensor < (MAX_BD_HEIGHT) - 0.1f
+          ) {
+            babystep.set_mm(Z_AXIS, cur_z - z_sensor);
+            DEBUG_ECHOLNPGM("BD:", z_sensor, ", Z:", cur_z, "|", current_position.z);
+          }
+          else {
+            babystep.set_mm(Z_AXIS, 0);   //if (old_cur_z <= cur_z) Z_DIR_WRITE(HIGH);
+            stepper.apply_directions();   // TODO: Remove this line as probably not needed
+          }
         }
       #endif
+
       old_cur_z = cur_z;
       old_buf_z = current_position.z;
       endstops.bdp_state_update(z_sensor <= 0.01f);
-      //endstops.update();
+
+      #if HAS_STATUS_MESSAGE
+        static float old_z_sensor = 0;
+        if (old_z_sensor != z_sensor) {
+          old_z_sensor = z_sensor;
+          char tmp_1[32];
+          sprintf_P(tmp_1, PSTR("BD:%d.%02dmm"), int(z_sensor), int(z_sensor * 100) % 100);
+          //SERIAL_ECHOLNPGM("Bed Dis:", z_sensor, "mm");
+          ui.set_status(tmp_1, true);
+        }
+      #endif
     }
     else
       stepper.apply_directions();
```

Comment:
What is the reason to call `stepper.apply_directions()` here and above? Such a hack should not be required. If you find that this is necessary with the current `bugfix-2.1.x` code, that would be a bug. Some changes to `BABYSTEPPING` were recently made to ensure that the DIR state will not be left in a bad state. (c6ea99f4)